### PR TITLE
GitHub actions best practices

### PR DESCRIPTION
**Pull request by nahteb**
This addresses some of the [guidance GDS has on using Github Actions](https://gds-way.cloudapps.digital/standards/source-code.html#using-github-actions-and-workflows), specifically:
- Pinning actions to a specific version
- Using a CODEOWNERS file to protect the `.github/workflows` folder
- Creating your own actions wherever possible

### CODE REVIEW

**Review by jamietanna — COMMENTED** [review]
Looking good! We've also got an alert (through Scorecards) that we need to specify a top-level permission for `.github/workflows/examples-java-codeql-analysis.yml` if you'd mind adding that? It should be as limited as possible, as the job itself gets more access

**Review by jamietanna — COMMENTED** [review]
Not sure I 100% agree with https://github.com/co-cddo/federated-api-model/pull/77/commits/7b76de1f16fc2dc7407e39e1be6787303e7e4da1 as it leads to a lot more overhead for us - thoughts?

File: .github/workflows/examples-java-push.yml [hunk 1]
```diff
@@ -65,20 +65,32 @@ jobs:
...
+      - uses: actions/download-artifact@f023be2c48cc18debc3bacd34cb396e0295e2869
         with:
           name: examples-java-libs
-      - uses: actions/download-artifact@v2
+      - uses: actions/download-artifact@f023be2c48cc18debc3bacd34cb396e0295e2869
         with:
           name: spring-boot-manifest
+      - name: Install the CF CLI
+        run: |
+          wget -q -O cf.tar.gz "https://packages.cloudfoundry.org/stable?release=linux64-binary&version=7.4.0&source=github-rel"
```
Comment (jamietanna) [inline .github/workflows/examples-java-push.yml:82]: it would be good to see if there's any other way of pulling this that may allow us to Dependabot it?

**Review by jamietanna — COMMENTED** [review]
Can `Add a CODEOWNERS file` explain why, please?

**Comment by nahteb** [comment]
> Not sure I 100% agree with [7b76de1](https://github.com/co-cddo/federated-api-model/commit/7b76de1f16fc2dc7407e39e1be6787303e7e4da1) as it leads to a lot more overhead for us - thoughts?

I think if we wanted to update to newer versions of CF CLI this action would be a blocker, especially as it hasn't been updated for 15 months

File: .github/workflows/examples-java-push.yml (hunk 1 above)
Comment (nahteb) [reply .github/workflows/examples-java-push.yml:82]: there might be, will have a look

Comment (nahteb) [reply .github/workflows/examples-java-push.yml:82]: had a look but can't see a nice way to do that unfortunately

File: .github/workflows/examples-java-push.yml [hunk 2]
```diff
@@ -65,20 +65,32 @@ jobs:
...
-        uses: citizen-of-planet-earth/cf-cli-action@v2
-        with:
-          cf_api: https://api.london.cloud.service.gov.uk
-          cf_username: ${{ secrets.CF_USER }}
-          cf_password: ${{ secrets.CF_PASSWORD }}
-          cf_org: cabinet-office-cddo-dsa-apis
-          cf_space: sandbox
-          command: push -f manifest.yml -p examples/java/spring-boot/build/libs/federated-api-model.jar
+        run: |
+          cf push -f manifest.yml --strategy rolling
```
Comment (jamietanna) [inline .github/workflows/examples-java-push.yml:95]: Can we add the rolling deployment as a separate commit as it closes https://github.com/co-cddo/federated-api-model/issues/59?

**Comment by jamietanna** [comment]
+ mind rebasing? :pray:

**Review by jamietanna — APPROVED** [review]

**Pull request merged**